src/Repository/Storable/Introduction.php

```diff
@@ -0,0 +1,155 @@
+ */
+class Introduction extends Storable
+{
+	protected static $table_name = 'intro';
+
+	protected static $model_class = Model\Introduction::class;
+
+	protected static $collection_class = Collection\Introductions::class;
+
+	/**
+	 * Confirms a follow request and sends a notic to the remote contact.
+	 *
+	 * @param Model\Introduction $Introduction
+	 * @param bool               $duplex       Is it a follow back?
+	 * @param bool|null          $hidden       Should this contact be hidden? null = no change
+	 * @return bool
+	 * @throws HTTPException\InternalServerErrorException
+	 * @throws HTTPException\NotFoundException
+	 * @throws \ImagickException
+	 */
+	public function confirm(Model\Introduction $Introduction, bool $duplex = false, bool $hidden = null)
+	{
+		$this->logger->info('Confirming follower', ['cid' => $this->{'contact-id'}]);
+
+		$contact = Model\Contact::selectFirst([], ['id' => $Introduction->{'contact-id'}, 'uid' => $Introduction->uid]);
+
+		if (!$contact) {
+			throw new HTTPException\NotFoundException('Contact record not found.');
+		}
+
+		$newRelation = $contact['rel'];
+		$writable = $contact['writable'];
+
+		if (!empty($contact['protocol'])) {
+			$protocol = $contact['protocol'];
+		} else {
+			$protocol = $contact['network'];
+		}
+
+		if ($protocol == Protocol::ACTIVITYPUB) {
+			ActivityPub\Transmitter::sendContactAccept($contact['url'], $contact['hub-verify'], $contact['uid']);
+		}
+
+		if (in_array($protocol, [Protocol::DIASPORA, Protocol::ACTIVITYPUB])) {
+			if ($duplex) {
+				$newRelation = Model\Contact::FRIEND;
+			} else {
+				$newRelation = Model\Contact::FOLLOWER;
+			}
+
+			if ($newRelation != Model\Contact::FOLLOWER) {
+				$writable = 1;
+			}
+		}
+
+		$fields = [
+			'name-date' => DateTimeFormat::utcNow(),
+			'uri-date'  => DateTimeFormat::utcNow(),
+			'blocked'   => false,
+			'pending'   => false,
+			'protocol'  => $protocol,
+			'writable'  => $writable,
+			'hidden'    => $hidden ?? $contact['hidden'],
+			'rel'       => $newRelation,
+		];
+		$this->dba->update('contact', $fields, ['id' => $contact['id']]);
+
+		array_merge($contact, $fields);
+
+		if ($newRelation == Model\Contact::FRIEND) {
+			if ($protocol == Protocol::DIASPORA) {
+				$ret = Diaspora::sendShare(Model\Contact::getById($contact['uid']), $contact);
+				$this->logger->info('share returns', ['return' => $ret]);
+			} elseif ($protocol == Protocol::ACTIVITYPUB) {
+				ActivityPub\Transmitter::sendActivity('Follow', $contact['url'], $contact['uid']);
+			}
+		}
+
+		return $this->delete($Introduction);
+	}
+
+
+	/**
+	 * Silently ignores the introduction, hides it from notifications and prevents the remote contact from submitting
+	 * additional follow requests.
+	 *
+	 * @param Model\Introduction $Introduction
+	 * @return bool
+	 * @throws \Exception
+	 */
+	public function ignore(Model\Introduction $Introduction)
```

Comment:
The question we've to ask ourselves: Do we want to save the state, we're updating to the persistence layer, also in our current model? Because "just" calling the a dba update doesn't update the current state of execution!

That's why I introduced the `IStorable` interface and the `StorableTrait` class, to automatically update the instance itself.

And we've to ask ourselves if we "really" want to update everytime we ignore a introduction .. Or just set it inside the `Introduction` model instance and call `save` at least.. So we would shift the db-calls to the end of the execution path ;-) ;-) ... I think I can now better describe what I try to do with `Entities` ;-) ;-)

So "ignoring" an `Introduction` in the Module class itself would be:
`$Introduction->ignore = true`
and at the end:
`DI::intro()->save($Introduction);` which knows what to do.

=> So we would be aware of every state-change of the instance during the execution..

Reply:
I understand what you mean, but I'm weary of deferred save calls. I'm afraid this will be easy to forget and hard to debug and I don't believe this is what we need at the moment.

Reply:
I see, yes that might be true, but could be handled through a `__deconstructor()` call, which at least will save any not-saved changes and log it as a warning ("you would have forgot saving this instance!")

But do we need updated instances after db-calls? If so, we should make them aware of it .. like using the "ignored" `Introduction` instance at a later stage, where we would need to call the latest state of the db again ..

I'm currently in a project where at some days, millions of data is sent and has to be processed in ~1-2 hours. There we needed to avoid DB interactions as much as possible. So we selected all necessary data at once, processed it inside the memory, and saved it back at the end of the execution. This was the fastest way :-) .. That's what I've in mind everytime we talk about our current Friendica design ^^

Reply:
> 
> I also really don't like traits when it could be replaced by direct inheritance.
>
Understandable :-) Maybe I'm using it because I miss the "multi-inheritance" feature in all the other languages where I already would have "really" needed it ^^

Reply:
I know what you mean, but we are currently running under the assumption that we have the freshest data when we call `Contact::getById` for example, and it would be risky to port the whole codebase towards a model that doesn't.

Also, in Friendica we have roughly three main call structures:
- GET calls: Mostly reads, should be light on write, no data integrity risk and no improvement from data save deferment.
- POST calls: Mostly atomic writes, then redirects to a GET call, no data integrity risk and no expected improvement from data save deferment.
- Worker tasks: Write and Read heavy, possible data integrity risks, no necessary performance improvement from data save deferment so far.

In your job, it does make sense but not as much for Friendica given the risks and the little expected benefits.

Reply:
> Understandable :-) Maybe I'm using it because I miss this "multi-inheritance" feature in all the other languages where I already would have "really" needed it ^^

Yeah, I've lived all my dev life without multi-inheritance. It may be useful and I'm certainly used to not use it, but when you have one trait and no parent class, it feels like you are using a jackhammer where a mallet is enough.

Reply:
fair enough .. Just have it in mind if Friendica gets more complex (or we get stormed by items/new users in Fediverse ;-) )

Reply:
Yes, the scaling issue is in the back of my mind, but we aren't there yet.